fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/jbatch/bulkdata/export/fast/ResourcePayloadReader.java

```diff
@@ -0,0 +1,683 @@
+ *
+ * The Java Batch framework separates the reader from the writer. On the face of it, separation of
+ * concerns seems like a good idea. However, because the fetching of the data is controlled by the
+ * persistence layer and the fact that we may need to end one multi-part upload and continue feeding
+ * a new multi-part upload, the separation makes things a lot harder and less readable - there is a
+ * lack of cohesion.
+ *
+ * To address this, we handle both reading and writing in this class, leaving the ItemWriter as a
+ * minimal stub. The framework is still leveraged for checkpointing.
+ *
+ * The {@link #readItem()} call simply returns an Object if there is more data to process. The
+ * {@link ResourceExportCheckpointAlgorithm} will always say we need a checkpoint, so the
+ * checkpoint will occur immediately after each read/write.
+ *
+ * The export is complete when {@link #readItem()} returns null.
+ *
+ * The export assumes that the persistence layer can efficiently scan forward based on the last_updated
+ * time of a resource. It is possible that multiple resources may share the same last_updated time, so
+ * this class has to track which resources have already been processed for a given timestamp. This is
+ * fairly easy, because the persistence layer must provide the data in order of last_updated. Because
+ * the number of resources for a given timestamp is probably very small, it is more efficient to track
+ * it this way than asking the persistence layer (e.g. JDBC) to sort. This can cause performance issues,
+ * because it negates the benefit of traversing an index in order.
+ */
+@Dependent
+public class ResourcePayloadReader extends AbstractItemReader {
+    private final static Logger logger = Logger.getLogger(ResourcePayloadReader.class.getName());
+    private static final String CLASS = ResourcePayloadReader.class.getName();
+
+    // S3 client API to IBM Cloud Object Storage
+    private AmazonS3 cosClient = null;
+
+    // TODO ???
+    boolean isSingleCosObject = false;
+
+    // TODO understand this
+    // The current position in the id-list we read back after
+    int currentOffset = 0;
+
+
+    // The handle to the persistence instance used to fetch the resources we want to export
+    FHIRPersistence fhirPersistence;
+
+    // The resource type class of the resources being exported by this instance (derived from the injected fhirResourceType value
+    Class<? extends Resource> resourceType;
+
+    /**
+     * Fhir tenant id.
+     */
+    @Inject
+    @BatchProperty(name = Constants.FHIR_TENANT)
+    String fhirTenant;
+
+    /**
+     * Fhir data store id.
+     */
+    @Inject
+    @BatchProperty(name = Constants.FHIR_DATASTORE_ID)
+    String fhirDatastoreId;
+
+    /**
+     * Fhir resource type to process.
+     */
+    @Inject
+    @BatchProperty(name = Constants.PARTITION_RESOURCE_TYPE)
+    String fhirResourceType;
+
+
+    // TODO ???
```

Comment:
```suggestion
    /**
     * If true, we set the ACL on the exported COS Object so that end users with knowledge of the export key can download them.
     * If false, we do not set the ACL and users must access the objects using other means and/or out-of-band credentials (e.g. using HMAC auth).
     * Set from fhir-server-config.json.
     */
```